minor-modes.lisp

```diff
@@ -0,0 +1,1113 @@
+               (if (not (or (symbolp bind-to)
+                            (stringp bind-to)
+                            (functionp bind-to)))
+                   (restart-case (error "Invalid binding ~A" bind-to)
+                     (bind-anyway ()
+                       :report "Bind the key binding regardless"
+                       (traverse-and-bind in-seq))
+                     (skip-binding ()
+                       :report "skip this binding"
+                       nil))
+                   (traverse-and-bind in-seq))))))
+      (cond ((null keymap)
+             topmap)
+            ((or (symbolp keymap)
+                 (stumpwm::kmap-p keymap))
+             keymap)
+            ((listp keymap)
+             (restart-case (mapc (lambda (keys)
+                                   (create-keymap-binding keys)
+                                   (setf curmap topmap))
+                                 keymap)
+               (abort-bindings ()
+                 :report "Return the keymap without binding further keys"
+                 topmap)
+               (abort-bindings* ()
+                 :report "Return an empty keymap"
+                 (stumpwm:make-sparse-keymap)))
+             topmap)
+            (t (restart-case
+                   (error "Function MAKE-MINOR-MODE-KEYMAP cant understand ~A"
+                          keymap)
+                 (use-empty-keymap ()
+                   :report "Use an empty keymap"
+                   (stumpwm:make-sparse-keymap))))))))
+
+(defun make-minor-mode-keymap (spec)
+  (generate-keymap spec))
+
+(defun make-minor-mode-top-map (top-map-spec root-map-spec)
+  "Create a top map for a minor mode based upon its TOP-MAP-SPEC and
+ROOT-MAP-SPEC."
+  (let ((top-map nil)
+        (root-map (if root-map-spec
+                      (make-minor-mode-keymap root-map-spec)
+                      (make-sparse-keymap))))
+    (fill-keymap top-map *escape-key* root-map)
+    (generate-keymap top-map-spec top-map)))
+
+
+(eval-when (:compile-toplevel :load-toplevel :execute)
+  (defun make-special-variable-name (mode name)
+    (intern (format nil "*~A-~A*" mode name)))
+  
+  (defun parse-minor-mode-options (options)
+    (let ((valid-options
+            '((:interactive            . 1)
+              (:scope                  . 1)
+              (:global                 . 1)
+              (:lighter-make-clickable . 1)
+              (:lighter                . 1)
+              (:expose-keymaps         . 1)
+              (:rebind                 . 1)
+              (:root-map               . 1)
+              (:top-map                . 1)
+              (:enable-when            . t)
+              (:make-hooks             . 1)
+              (:default-initargs       . t)
+              (:define-command-definer . 1)))
+          (all-vals '())
+          (other-opts '()))
+      (flet ((collect-values (option)
+               (destructuring-bind (option-name &rest option-arguments) option
```

Comment:
```suggestion
               (destructuring-bind (option-name . option-arguments) option
```
We can use the cons literal notation here, no need &rest

Reply:
done and done :smile: